### PR TITLE
Code evaluator refactor

### PR DESCRIPTION
**Pull request by ankitjavalkar**

### CODE REVIEW

File: yaksh/bash_code_evaluator.py [hunk 1]
```diff
@@ -61,23 +46,31 @@ def _check_code(self, ref_path, submit_path,
...
-            return False, "No file at %s or Incorrect path" % submit_path
-        if not os.access(ref_path, os.X_OK):
-            return False, "Script %s is not executable" % ref_path
-        if not os.access(submit_path, os.X_OK):
-            return False, "Script %s is not executable" % submit_path
+        ref_code_path = test_case
+        get_ref_path, get_test_case_path = ref_code_path.strip().split(',')
+        get_ref_path = get_ref_path.strip()
+        get_test_case_path = get_test_case_path.strip()
+        clean_ref_code_path, clean_test_case_path = self._set_test_code_file_path(get_ref_path,
```
Comment (prabhuramachandran) [inline yaksh/bash_code_evaluator.py]: 80 cols?

File: yaksh/bash_code_evaluator.py [hunk 2]
```diff
@@ -88,26 +81,26 @@ def _check_code(self, ref_path, submit_path,
...
                 return False, err
         else:
-            if not isfile(test_case_path):
-                return False, "No test case at %s" % test_case_path
-            if not os.access(ref_path, os.R_OK):
-                return False, "Test script %s, not readable" % test_case_path
+            if not isfile(clean_test_case_path):
+                return False, "No test case at %s" % clean_test_case_path
+            if not os.access(clean_ref_code_path, os.R_OK):
+                return False, "Test script %s, not readable" % clean_test_case_path
```
Comment (prabhuramachandran) [inline yaksh/bash_code_evaluator.py]: 80 cols?

File: yaksh/code_evaluator.py [hunk 3]
```diff
@@ -98,54 +77,67 @@ def evaluate(self):
...
         result = {'success': success, 'error': err}
         return result
 
     # Private Protocol ##########
-    def _setup(self):
+    def setup(self):
         self._change_dir(self.in_dir)
 
-    def _evaluate(self, args):
+    def safe_evaluate(self, user_answer, test_case_data):
```
Comment (prabhuramachandran) [inline yaksh/code_evaluator.py]: Docstring for method please.

File: yaksh/cpp_code_evaluator.py [hunk 4]
```diff
@@ -12,114 +12,122 @@
...
+            self.compiled_test_code = self._run_command(self.compile_main,
+                                     shell=True,
+                                     stdout=subprocess.PIPE,
+                                     stderr=subprocess.PIPE)
+
+
+            return self.compiled_user_answer, self.compiled_test_code
+
+    def check_code(self, user_answer, test_case):
+            """ Function validates student code using instructor code as
```
Comment (prabhuramachandran) [inline yaksh/cpp_code_evaluator.py]: Messed up indentation.

File: yaksh/java_code_evaluator.py [hunk 5]
```diff
@@ -12,48 +12,63 @@
...
+
+    def compile_code(self, user_answer, test_case):
+        if self.compiled_user_answer and self.compiled_test_code:
+            return None
+        else:
+            ref_code_path = test_case
+            clean_ref_code_path, clean_test_case_path = self._set_test_code_file_path(ref_code_path)
+
+            if not isfile(clean_ref_code_path):
+                return False, "No file at %s or Incorrect path" % clean_ref_code_path
```
Comment (prabhuramachandran) [inline yaksh/java_code_evaluator.py]: 80 cols?

File: yaksh/models.py [hunk 6]
```diff
@@ -231,6 +216,18 @@ def load_from_json(self, questions_list, user):
             question['user'] = user
             Question.objects.get_or_create(**question)
 
+    def get_test_cases(self, **kwargs):
+        test_case_ctype = ContentType.objects.get(app_label="yaksh", model=self.test_case_type)
```
Comment (prabhuramachandran) [inline yaksh/models.py]: 80 cols?

File: yaksh/tests.py [hunk 7]
```diff
@@ -507,3 +466,48 @@ def test_is_teacher(self):
         self.course.add_teachers(self.student2)
         result = self.course.is_teacher(self.student2)
         self.assertTrue(result)
+
+
+###############################################################################
+class TestCaseTestCases(unittest.TestCase):
+    def setUp(self):
+        self.user = User.objects.get(pk=1)
+        self.question1 = Question(summary='Demo question 1', language='Python',
```
Comment (prabhuramachandran) [inline yaksh/tests.py]: 80 cols?

File: yaksh/bash_code_evaluator.py [hunk 8]
```diff
@@ -119,4 +128,3 @@ def _check_code(self, ref_path, submit_path,
                 err = "Error:expected %s, got %s" % (inst_stdout+inst_stderr,
```
Comment (ankitjavalkar) [inline yaksh/bash_code_evaluator.py]: line wrapped

File: yaksh/evaluator_tests/test_c_cpp_evaluation.py [hunk 9]
```diff
@@ -5,71 +5,49 @@
 
 class CEvaluationTestCases(unittest.TestCase):
     def setUp(self):
-        self.language = "C"
-        self.ref_code_path = "c_cpp_files/main.cpp"
+        self.test_case_data = [{"test_case": "c_cpp_files/main.cpp"}]
         self.in_dir = "/tmp"
-        self.test_case_data = []
         self.timeout_msg = ("Code took more than {0} seconds to run. "
                             "You probably have an infinite loop in your code.").format(SERVER_TIMEOUT)
```
Comment (ankitjavalkar) [inline yaksh/evaluator_tests/test_c_cpp_evaluation.py]: line wrapped

File: yaksh/evaluator_tests/test_code_evaluation.py [hunk 10]
```diff
@@ -1,20 +1,21 @@
...
     def setUp(self):
-        set_registry()
         self.registry_object = get_registry()
         self.language_registry = _LanguageRegistry()
 
     def test_set_register(self):
-        class_name = getattr(python_code_evaluator, 'PythonCodeEvaluator')
-        self.registry_object.register("python", "yaksh.python_code_evaluator.PythonCodeEvaluator")
-        self.assertEquals(self.registry_object.get_class("python"), class_name)
+        class_name = getattr(python_assertion_evaluator, 'PythonAssertionEvaluator')
```
Comment (ankitjavalkar) [inline yaksh/evaluator_tests/test_code_evaluation.py]: line wrapping

File: yaksh/evaluator_tests/test_code_evaluation.py [hunk 11]
```diff
@@ -1,20 +1,21 @@
...
+        self.registry_object.register("python", {"standardtestcase": "yaksh.python_assertion_evaluator.PythonAssertionEvaluator",
+                    "stdoutbasedtestcase": "python_stdout_evaluator.PythonStdoutEvaluator"
+            })
+        self.assertEquals(self.registry_object.get_class("python", "standardtestcase"), class_name)
```
Comment (ankitjavalkar) [inline yaksh/evaluator_tests/test_code_evaluation.py]: line wrapping

File: yaksh/evaluator_tests/test_java_evaluation.py [hunk 12]
```diff
@@ -2,52 +2,61 @@
...
 
 class JavaEvaluationTestCases(unittest.TestCase):
     def setUp(self):
-        self.language = "java"
-        self.ref_code_path = "java_files/main_square.java"
+        self.test_case_data = [{"test_case": "java_files/main_square.java"}]
         self.in_dir = "/tmp"
-        self.test_case_data = []
         evaluator.SERVER_TIMEOUT = 9
         self.timeout_msg = ("Code took more than {0} seconds to run. "
```
Comment (ankitjavalkar) [inline yaksh/evaluator_tests/test_java_evaluation.py]: line wrapping

File: yaksh/evaluator_tests/test_java_evaluation.py [hunk 13]
```diff
@@ -2,52 +2,61 @@
...
 from yaksh import code_evaluator as evaluator
 from yaksh.java_code_evaluator import JavaCodeEvaluator
+from yaksh.settings import SERVER_TIMEOUT
 
 
 class JavaEvaluationTestCases(unittest.TestCase):
     def setUp(self):
-        self.language = "java"
-        self.ref_code_path = "java_files/main_square.java"
+        self.test_case_data = [{"test_case": "java_files/main_square.java"}]
```
Comment (ankitjavalkar) [inline yaksh/evaluator_tests/test_java_evaluation.py]: line wrapping

File: yaksh/evaluator_tests/test_python_evaluation.py [hunk 14]
```diff
@@ -134,14 +137,64 @@ def add(a, b):
...
         self.assertFalse(result.get("success"))
         self.assertEqual(2, len(err))
         for msg in value_error_msg:
             self.assertIn(msg, result.get("error"))
 
+class PythonStdoutEvaluationTestCases(unittest.TestCase):
+    def setUp(self):
+        self.test_case_data = [{"expected_output": "0 1 1 2 3"}]
+        self.timeout_msg = ("Code took more than {0} seconds to run. "
+                            "You probably have an infinite loop in your code.").format(SERVER_TIMEOUT)
```
Comment (ankitjavalkar) [inline yaksh/evaluator_tests/test_python_evaluation.py]: line wrapping

File: yaksh/java_code_evaluator.py [hunk 15]
```diff
@@ -12,48 +12,77 @@
...
+    def compile_code(self, user_answer, test_case):
+        if self.compiled_user_answer and self.compiled_test_code:
+            return None
+        else:
+            ref_code_path = test_case
+            clean_ref_code_path, clean_test_case_path = \
+                     self._set_test_code_file_path(ref_code_path)
+
+            if not isfile(clean_ref_code_path):
+                msg = "No file at %s or Incorrect path" % clean_ref_code_path
```
Comment (ankitjavalkar) [inline yaksh/java_code_evaluator.py]: line wrapping

File: yaksh/java_code_evaluator.py [hunk 16]
```diff
@@ -12,48 +12,77 @@
...
+                return False, msg
+            if not isfile(self.submit_code_path):
+                msg = "No file at %s or Incorrect path" % self.submit_code_path
```
Comment (ankitjavalkar) [inline yaksh/java_code_evaluator.py]: line wrapping

File: yaksh/java_code_evaluator.py [hunk 17]
```diff
@@ -12,48 +12,77 @@
...
+                return False, msg
+
+            user_code_directory = os.getcwd() + '/'
+            self.write_to_submit_code_file(self.submit_code_path, user_answer)
```
Comment (ankitjavalkar) [inline yaksh/java_code_evaluator.py]: line wrapping

File: yaksh/java_code_evaluator.py [hunk 18]
```diff
@@ -12,48 +12,77 @@
...
+            ref_file_name = (clean_ref_code_path.split('/')[-1]).split('.')[0]
+            self.user_output_path = self.set_file_paths(user_code_directory,
```
Comment (ankitjavalkar) [inline yaksh/java_code_evaluator.py]: line wrappings

**Pull request merged**